internal/cmd/ssh.go

```diff
@@ -167,72 +171,125 @@ func setupDestinationSSHConfig(ctx context.Context, cli *CLI, destination *api.D
 		return err
 	}
 
-	if err := provisionSSHKey(ctx, cli, client, infraSSHDir); err != nil {
+	user, err := client.GetUserSelf(ctx)
+	if err != nil {
+		return err
+	}
+
+	keyFilename, err := provisionSSHKey(ctx, provisionSSHKeyOptions{
+		cli:         cli,
+		client:      client,
+		hostConfig:  hostCfg,
+		infraSSHDir: infraSSHDir,
+		user:        user,
+	})
+	if err != nil {
 		return fmt.Errorf("create ssh keypair: %w", err)
 	}
 
 	if err := writeInfraKnownHosts(infraSSHDir, destination); err != nil {
 		return fmt.Errorf("write known hosts: %w", err)
 	}
 
-	user, err := client.GetUserSelf(ctx)
-	if err != nil {
-		return err
-	}
-
-	if err := writeDestinationSSHConfig(infraSSHDir, destination, user); err != nil {
+	if err := writeDestinationSSHConfig(infraSSHDir, destination, user, keyFilename); err != nil {
 		return fmt.Errorf("write infra ssh config: %w", err)
 	}
 	return nil
 }
 
-func provisionSSHKey(ctx context.Context, cli *CLI, client *api.Client, infraSSHDir string) error {
-	keyFilename := filepath.Join(infraSSHDir, "key")
+type provisionSSHKeyOptions struct {
+	cli         *CLI
+	client      *api.Client
+	infraSSHDir string
+	hostConfig  *ClientHostConfig
+	user        *api.User
+}
+
+func provisionSSHKey(ctx context.Context, opts provisionSSHKeyOptions) (string, error) {
+	keysCfg, err := readKeysConfig(opts.infraSSHDir)
+	switch {
+	case errors.Is(err, fs.ErrNotExist):
+		keysCfg = &keysConfig{}
+	case err != nil:
+		return "", err
+	}
 
-	// TODO: check expiration
-	// TODO: check the key exists in the API
-	if fileExists(keyFilename) {
-		return nil
+	org, err := opts.client.GetOrganizationSelf(ctx)
+	if err != nil {
+		return "", err
 	}
 
-	fmt.Fprintf(cli.Stderr, "Creating a new RSA 4096 bit key pair in %v\n", keyFilename)
+	keysDir := filepath.Join(opts.infraSSHDir, "keys")
+	existingKeys := matchingPublicKeys(keysCfg, opts.hostConfig, org.ID)
+	for _, existing := range existingKeys {
+		filename := filepath.Join(keysDir, existing.PublicKeyID)
+		if !fileExists(filename) || !fileExists(filename+".pub") {
+			// key doesn't exist locally
+			continue
+		}
+
+		if userPublicKeyContains(opts.user.PublicKeys, existing.PublicKeyID) {
+			// TODO: check expiration when expiry is added
+			// key exists locally and in the API
+			return filename, nil
+		}
+
+		// key doesn't exist in the API
+		// TODO: delete the local key file, and remove it from keysConfig
+	}
+
+	_ = os.MkdirAll(keysDir, 0o700)
+	fmt.Fprintf(opts.cli.Stderr, "Creating a new RSA 4096 bit key pair in %v\n", keysDir)
```

Comment:
Does the user need to know everytime this happens? Feels like it could just happen behind the scenes.

Reply:
I think it's important we tell the user what is happening for a few reasons.  The main one is that generating a new key is relatively slow. It can easily take a second or even two. If we don't tell the user what is happening then they get the impression that Infra is slow or sluggish.  By giving them information about what is happening , they can better understand why some runs of `ssh` are fast, and others appear slow.

In general I think it's always better to give the user information, than to hide it. It's definitely possible to go too far and spew tons of logs, but here it's a single informative log message. If a user ever has to debug a problem with `infra`, they will be in a better position because we've given them some basic information about how it works. If we don't do this users are left feeling helpless and that's also a bad experience.